### PR TITLE
Observability Helper Library

### LINKED ISSUE #17: AML Observability Library
Requirements:

1. The model scoring application should be able to log metrics and trace events to App Insights.
2. Will also be to nice to have the same code to log metrics to both App Insights and Azure ML during training.

### PR DESCRIPTION
**Pull request by h2floh**
Implements: #17 

Started from MLOpsPython [Observability PR at commit](https://github.com/microsoft/MLOpsPython/pull/307/commits/c11342302123bf7a812e560726a3a509b937d13b) and made some smaller modifications:

* Added [tracing features](https://pypi.org/project/opencensus-ext-azure/) for logs and http calls
* Added a "Console Logger" for local debugging/development
* Added exception logging
* Added `custom properties` to get information about the AML run into the [logs/traces/exceptions](https://docs.microsoft.com/en-us/azure/azure-monitor/app/opencensus-python#logs)
* Added `sampling rates` and `log level` as env variables
* Moved namespace to `ml_service` since it anyway depends on `ml_service.util.env_variables`

> **IMPORTANT**: In order for the pipeline step scripts to be able to use the logger, the full repo needs to be made available, so the TRAIN_SCRIPT_PATH is now set to repo root `./` and all script variables need to be referenced from there e.g. `ml_model/preprocessing/process_aml.py`

Disclaimer
* Not yet worked on or extended the unit tests
* Not yet tested ML metrics logging to AML

### CODE REVIEW

**Review by tarockey — APPROVED** [review]
This looks great! I really like the integration of the 3 different logging options. 
One thing to consider (or maybe I missed it), would be to add the ability to specify which logger a singular message goes. For example, I generally want to log all metrics to both Azure ML and App Insights, but maybe I just want to send exceptions to App Insights and not AzureML.

Also, be careful with having the TRAIN_SCRIPT_ROOT being set to `./`, as that means that every file is uploaded to AzureML for each run - it is worth digging into .amlignore, or looking into another route for accessing the logger (maybe publishing a wheel?)

**Comment by h2floh** [comment]
Hey @tarockey 
> One thing to consider (or maybe I missed it), would be to add the ability to specify which logger a singular message goes. For example, I generally want to log all metrics to both Azure ML and App Insights, but maybe I just want to send exceptions to App Insights and not AzureML.

Mainly the Azure ML logger is just a  "print to console" for .log and .exception, the only difference is that log_metric will use the mechanism to record AML metrics directly within AMLS on the run. This feature is more or less as is copied from the MLOpsPython PR.

You probably want to have the exceptions logged in the AML log text files and in AppInsights. But I am open to add some flags.

> 
> Also, be careful with having the TRAIN_SCRIPT_ROOT being set to `./`, as that means that every file is uploaded to AzureML for each run - it is worth digging into .amlignore, or looking into another route for accessing the logger (maybe publishing a wheel?)

Yeah, thanks that you also found that :) This is the main drawback and we are already considering packing the observability lib. Therefore we probably won't merge this PR but first refactor the repo structure into folders for different examples and tools. The planned structure is:

- samples 
  - image-classification-tensorflow
     - .pipelines
     - docs
     - model
     - ml_service
     - ...
  - object-detection-yolo 
  - non-pyton
  - ...
- common
  - **observability-lib** <-- create a package which can be imported by the samples
  - docs
     - variable management
     - security best practice
--- ...
- readme

File: ml_service/util/logger/app_insights_logger.py [hunk 1]
```diff
@@ -0,0 +1,188 @@
...
+
+class AppInsightsLogger(LoggerInterface, ObservabilityAbstract):
+    def __init__(self, run):
+        print('Initializing the AppInsightsLogger')
+        self.env = Env()
+        self.run_id = self.get_run_id_and_set_context(run)
+
+        # Prepare integrations and initialize tracer
+        config_integration.trace_integrations(['httplib', 'logging'])
+        texporter = AzureExporter(connection_string=self.
```
Comment (kenakamu) [inline ml_service/util/logger/app_insights_logger.py:34]: Is there any reason to separate lines here?

File: ml_service/util/logger/app_insights_logger.py [hunk 2]
```diff
@@ -0,0 +1,188 @@
...
+        texporter.add_telemetry_processor(self.callback_function)
+        self.tracer = Tracer(
+            exporter=texporter,
+            sampler=ProbabilitySampler(self.env.trace_sampling_rate)
+        )
+
+        # Create AppInsights Handler and set log format
+        self.logger = logging.getLogger(__name__)
+        self.logger.setLevel(
+            getattr(logging, self.env.log_level.upper(), "WARNING"))
```
Comment (kenakamu) [inline ml_service/util/logger/app_insights_logger.py:45]: Just curious why default is set to WARNING when we set DEBUG/INFO in env? Do we want to set the lowest level instead?

File: ml_service/util/logger/logger_interface.py [hunk 3]
```diff
@@ -0,0 +1,152 @@
...
+
+    def get_run_id_and_set_context(self, run):
+        """
+        gets the correlation ID by the in following order:
+        - If the script is running  in an Online run Context of AML --> run_id
+        - If the script is running where a build_id
+        environment variable  is set --> build_id
+        - Else --> generate a unique id
+
+        Sets also the custom context dimensions based on On or Offline run
```
Comment (kenakamu) [inline ml_service/util/logger/logger_interface.py:86]: on Online or Offline?

File: ml_service/util/logger/logger_interface.py [hunk 4]
```diff
@@ -0,0 +1,152 @@
...
+        :param run:
+        :return: correlation_id
+        """
+        run_id = str(uuid.uuid1())
+        if not run.id.startswith(self.OFFLINE_RUN):
+            run_id = run.id
+            self.custom_dimensions = {
+                'custom_dimensions': {
```
Comment (kenakamu) [inline ml_service/util/logger/logger_interface.py:94]: Adding run number is useful when display the result as guid is not easy to identify which run it is.
Also adding workspace is useful when we use this module in multiple workspace but logging into same appinsights.

File: .pipelines/variables-template.yml [hunk 5]
```diff
@@ -42,6 +42,8 @@ variables:
     value: flower_custom_preprocess_env
 
   # AML Compute Cluster Config
+  - name: AML_ENV_TRAIN_CONDA_DEP_FILE
```
Comment (liupeirong) [inline .pipelines/variables-template.yml:45]: Is this var used?

File: .pipelines/variables-template.yml [hunk 6]
```diff
@@ -32,8 +32,8 @@ variables:
   - name: ALLOW_RUN_CANCEL
     value: "false"
   # Flag to allow rebuilding the AML Environment after it was built for the first time. This enables dependency updates from conda_dependencies.yaml.
-  # - name: AML_REBUILD_ENVIRONMENT
-  #  value: "false"
+  - name: AML_REBUILD_ENVIRONMENT
+    value: "true"
```
Comment (liupeirong) [inline .pipelines/variables-template.yml:36]: We only want to rebuild the first time the dependencies are changed, right? Perhaps leave the default to false, and set the ADO variable to true?

File: ml_model/preprocessing/Dockerfile [hunk 7]
```diff
@@ -25,8 +25,18 @@ RUN wget --quiet https://repo.anaconda.com/miniconda/Miniconda3-${CONDA_VERSION}
...
+RUN apt-get update --fix-missing && \
+    apt-get install -y build-essential
+
+# Create conda environment for dockeruser user
+USER dockeruser
+
 RUN conda install -y conda=${CONDA_VERSION} python=${PYTHON_VERSION} && \
     pip install azureml-defaults==${AZUREML_SDK_VERSION} inference-schema==${INFERENCE_SCHEMA_VERSION} &&\
+    pip install python-dotenv==0.12.* dataclasses==0.6 opencensus==0.7.11 opencensus-ext-httplib==0.7.3 \
+    opencensus-ext-azure==1.0.5 opencensus-ext-logging==0.1.0 opencensus-context==0.1.2 && \
```
Comment (liupeirong) [inline ml_model/preprocessing/Dockerfile:39]: should we use conda dependency yml to be consistent rather than directly pip install?

**Comment by liupeirong** [comment]
Let's pack it to a lib rather than having all samples repeat the observability code, as noted, it'll also keep the training folder smaller.

**Comment by h2floh** [comment]
> lib rather than having all samples re

Yes, didn't had the chance to close this PR. It is outdated because we did the folder structure refactoring. We have to decide about things like account creation and management for the pip package repo. How it should be named etc. 

After that we reincorporate the logging methods into the existing samples, this should be more or less copy&paste work based on the work done here. Also good for splitting up the work/PRs for each sample. 

Thanks for any comments let's incorporate them into the refactored observability lib. I'll now close this PR for now.

@liupeirong @kenakamu

**Comment by h2floh** [comment]
Completed with #59

**Pull request closed without merging**